lib/matplotlib/text.py

```diff
@@ -66,15 +66,26 @@ def _get_text_metrics_with_cache(renderer, text, fontprop, ismath, dpi):
     """Call ``renderer.get_text_width_height_descent``, caching the results."""
     # Cached based on a copy of fontprop so that later in-place mutations of
     # the passed-in argument do not mess up the cache.
-    return _get_text_metrics_with_cache_impl(
-        weakref.ref(renderer), text, fontprop.copy(), ismath, dpi)
-
-
-@functools.lru_cache(4096)
-def _get_text_metrics_with_cache_impl(
-        renderer_ref, text, fontprop, ismath, dpi):
-    # dpi is unused, but participates in cache invalidation (via the renderer).
-    return renderer_ref().get_text_width_height_descent(text, fontprop, ismath)
+    func = _get_text_metrics_function(renderer)
+    return func(text, fontprop.copy(), ismath, dpi)
+
+
+# use mutable default to carry hidden global state
+def _get_text_metrics_function(inp_renderer, _cache=weakref.WeakKeyDictionary()):
+    if (_text_metrics := _cache.get(inp_renderer, None)) is None:
+        renderer_ref = weakref.ref(inp_renderer)
```

Comment:
Do you actually need to explicitly create a weakref here?  Doesn't using a WeakKeyDictionary basically do this for you for free already (i.e. aren't you having two layers of weakref'ing here?)?  (not sure...)

Reply:
yes, if we do not the closure holds a hard reference mediated in the value side of the weakkey dict and they become immortal.

Reply:
Ah yes, I missed the fact that you are caching a closure.

Reply:
Perhaps the closure can be made more explicit, something like (untested)
```python
cache = WeakKeyDictionary()  # Can be hidden as an attribute or a private parameter...


def _get_text_metrics_with_cache(renderer, text, fp, ismath, dpi):
    if renderer not in cache:
        cache[renderer] = functools.lru_cache(4096)(
            functools.partial(_weak_text_metrics, weakref.ref(renderer)))
    return cache[renderer](text, fp.copy(), ismath, dpi)


def _weak_text_metrics(renderer_ref, text, fp, ismath, dpi):
    return renderer_ref().get_text_width_height_descent(text, fp, ismath)
```

Reply:
I added a some comments.  I think that the closure is marginally clearer than the use of `partial`.